Implement tdd

kgebregewergs-epam opened the pull request:
## Summary of change

I added three test files which follows the TDD test step guild line. 

- TemplateEngineTest
- MailServerTest
- MessengerTest

kgebregewergs-epam committed:
Added logical unit to which are  all red or the first stage of TDD for TemplateEngineTest.

kgebregewergs-epam committed:
Fixed the templateEngineTest logical unit test to make all the unit tests pass.
Added mockito framework to mock the interfaces I created (fileInput and ConsoleInput).

kgebregewergs-epam committed:
Added the initial unit test for the business logic of MailServerTest.

kgebregewergs-epam committed:
added the second step to MailServerTest.

kgebregewergs-epam committed:
Added the first step for messengerTest.
in this step I added the business logic unit test with each implementation of failed or Red step of TDD.

kgebregewergs-epam committed:
Added the second step for messengerTest. all the tests are green at this time.

kgebregewergs-epam committed:
Refactored the messengerTest.

added a minor fix for mailserver class in order to follow the code check guild line.

QwerTech commented on src/test/java/com/epam/ld/module2/testing/template/TemplateEngineTest.java:
```diff
@@ -1,4 +1,38 @@
+
+    @Test
+    public void checkGenerateMessageWithValidTemplateFromConsole() {
+        fail();
```
Usually when we write test with TDD we actually implement tests fully, but the implementation of the production classes is missing.

QwerTech commented on src/main/java/com/epam/ld/module2/testing/MailServer.java:
```diff
@@ -4,13 +4,24 @@
      */
-    public void send(String addresses, String messageContent) {
+    public Boolean send(String addresses, String messageContent) {
+        if (addresses == null && messageContent == null) {
```
(minor) I think the best practice is to check and throw per each input parameter.

QwerTech commented on src/main/java/com/epam/ld/module2/testing/MailServer.java:
```diff
@@ -4,13 +4,24 @@
+        }
+
+        if (addresses != null && messageContent != null) {
+            System.out.format("Sending the message Successfully message:  ", messageContent);
```
(minor) In real applications we tend to use logging libraries instead of direct `System.out`

QwerTech commented on src/main/java/com/epam/ld/module2/testing/Messenger.java:
```diff
@@ -28,10 +28,21 @@ public Messenger(MailServer mailServer,
+        }
+
+        else if(client == null) {
+            return null;
```
Do we really need to return something here? It isn't used anywhere, except tests. 
Besides returning null from methods is usually a bad idea

kgebregewergs-epam replied:
Refactored this

QwerTech commented on src/main/java/com/epam/ld/module2/testing/Messenger.java:
```diff
@@ -28,10 +28,21 @@ public Messenger(MailServer mailServer,
-            templateEngine.generateMessage(template, client);
-        mailServer.send(client.getAddresses(), messageContent);
+    public String sendMessage(Client client, Template template) {
+        if(client != null && template != null) {
```
(minor) Check the article https://refactoring.guru/replace-nested-conditional-with-guard-clauses
It helps in such cases

QwerTech commented on src/test/java/com/epam/ld/module2/testing/MailServerTest.java:
```diff
@@ -1,4 +1,33 @@
+    @Test
+    public void shouldThrowNullPointerExecution() throws NullPointerException {
+        MailServer mailServer = new MailServer();
+        try {
```
There is `org.junit.jupiter.api.Assertions#assertThrows(java.lang.Class<T>, org.junit.jupiter.api.function.Executable)` specifically for such cases.

QwerTech commented on src/test/java/com/epam/ld/module2/testing/MailServerTest.java:
```diff
@@ -1,4 +1,33 @@
+            Boolean result  = mailServer.send(null, null);
+            fail();
+        } catch (NullPointerException nullPointerException) {
+            System.out.println("Null pointer Exception: " + nullPointerException.getMessage());
```
Printing something in tests is rarely a good idea, as tests suppose to follow KISS and should have explaining name and assertion which provides good explanation what went wrong.

QwerTech commented on src/test/java/com/epam/ld/module2/testing/MessengerTest.java:
```diff
@@ -0,0 +1,55 @@
+
+public class MessengerTest {
+
+    @Spy
```
Using annotation and then initializing it with spy(...) method call is confusing

QwerTech commented on src/test/java/com/epam/ld/module2/testing/MessengerTest.java:
```diff
@@ -0,0 +1,55 @@
+    public  void checkSendMessageShouldFailWhenClientIsNull() {
+        doReturn(null).when(templateEngine).generateMessage(template, null);
+        String result = messenger.sendMessage(null, template);
+        assertFalse(result != null);
```
There is `org.junit.jupiter.api.Assertions#assertNull(java.lang.Object)` for such cases

QwerTech commented on src/main/java/com/epam/ld/module2/testing/template/TemplateEngine.java:
```diff
@@ -2,18 +2,47 @@
  * The type Template engine.
  */
 public class TemplateEngine {
+    Set<String> validTemplates = new HashSet<>(Arrays.asList("${subject}", "${body}", "${cc}"));
```
(minor) In newer versions of Java there is `Set.of(...)`

QwerTech commented on src/main/java/com/epam/ld/module2/testing/template/TemplateEngine.java:
```diff
@@ -2,18 +2,47 @@
+     * @param template the template
+     * @return boolean
+     */
+    private boolean checkTemplate(Template template) {
```
It's a bit inconsistent that sometimes we throw exception and in other cases use boolean.
Java way is throwing exceptions.

QwerTech commented on src/test/java/com/epam/ld/module2/testing/MailServerTest.java:
```diff
@@ -1,4 +1,33 @@
 public class MailServerTest {
+
+    @Test
+    public void checkSuccessfulSendMessage() {
```
Tests names should be a bit more specific, like `send_shouldSendSuccessfully_whenPassedValidAddressAndTemplate`

QwerTech commented on src/test/java/com/epam/ld/module2/testing/MailServerTest.java:
```diff
@@ -1,4 +1,33 @@
+    }
+
+    @Test
+    public void checkUnSuccessfulSendingMessage() {
```
In that case, I would call method `send_shouldReturnFalse_whenContentIsNull`

QwerTech commented on src/test/java/com/epam/ld/module2/testing/MessengerTest.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Test
+    public  void checkSendMessageShouldFailWhenTemplateEngineIsNull() {
```
This is a good test name!

QwerTech commented on src/test/java/com/epam/ld/module2/testing/template/TemplateEngineTest.java:
```diff
@@ -1,4 +1,170 @@
 public class TemplateEngineTest {
+
+    @Test
+    public void checkGenerateMessageWithValidTemplateFromConsole() {
```
Isn't very clear why are we creating `ConsoleInput` and then copying data from it.
Also, seems like classes `ConsoleInput` `FileInput` `Input` are not used in production code.

QwerTech reviewed (changes requested):
- Add an app class which could be executed to show full functionality.
- That requirement looks like not implemented. Expected to see it in `TemplateEngine`
> The system replaces variable placeholders like #{subject} from a template with values provided at runtime. 


1. TDD approach – (2 points); **// 1.5 points as tests was empty in the first commit**
2. @Parameterized runner and Dynamic tests (0.3 points); **// missing**
3. Implement meta-annotations and filtering (0.3 points); **// missing**
4. TemporaryFolder rule (0.3 points); **// missing**
5. Mock reading from file/console (0.3 points); **// missing**
6. Use partial mock (0.3 points); **// 0.3 points**
7. Use spy (0.3 points); **// 0.3 points**
8. Create custom extension (jUnit5) to output test execution information to file (0.3 points); **// missing**
9. Using ExpectedException rule to check exceptions + Assertion mechanism (0.3 points); **// missing**
10. Implement Disabling test on condition (0.1 points); **// missing**
11. Test quality and adequate coverage will be assessed as (0.5 points).**// 0.3 points. Coverage is fine, but tests quality as a bit low**

kgebregewergs-epam committed:
Added log4j dependency for better utilization of logs.

kgebregewergs-epam committed:
Added a helper class that will hide the implementation from the main app.

kgebregewergs-epam committed:
Refactor the tests based on the comments provided and also to have a better log information.

kgebregewergs-epam committed:
Refactor Client and Template data types to have a constructor with multiple and no arg constructor.

kgebregewergs-epam committed:
Refactor The actual implementation of the logics.

kgebregewergs-epam committed:
Added the main App that receive inputs from either console or file.
added a sample file for test.
and also I added a readme file.

kgebregewergs-epam committed:
Refactor Messenger class and messenger Test class with better best practice.

kgebregewergs-epam committed:
Refactor Messenger to throw null pointer exception when either of the argument are null

kgebregewergs-epam committed:
Add a java doc for better readability an documentation.

kgebregewergs-epam committed:
Removed the client object from TemplateEngine
Removed the console input and file input mocks from template engine to APP tests.
added an app unit tests.
minor refactor on the remaining tests.

QwerTech commented on src/test/java/com/epam/ld/module2/testing/MailServerTest.java:
```diff
@@ -1,4 +1,51 @@
+    }
+
+    @ParameterizedTest
+    @NullSource
```
I have to ask you to use any other params source, as that one is too simple and isn't really data test.
`@MethodSource` would be perfect.

QwerTech reviewed (changes requested):
- That requirement looks like not implemented. Expected to see it in `TemplateEngine`
> The system replaces variable placeholders like #{subject} from a template with values provided at runtime. 


1. TDD approach – (2 points); **// 1.5 points as tests was empty in the first commit**
2. @Parameterized runner and Dynamic tests (0.3 points); **// 0.15 points, better parametrized is required and dynamic tests are missing**
3. Implement meta-annotations and filtering (0.3 points); **// missing**
4. TemporaryFolder rule (0.3 points); **// missing**
5. Mock reading from file/console (0.3 points); **// missing**
6. Use partial mock (0.3 points); **// 0.3 points**
7. Use spy (0.3 points); **// 0.3 points**
8. Create custom extension (jUnit5) to output test execution information to file (0.3 points); **// missing**
9. Using ExpectedException rule to check exceptions + Assertion mechanism (0.3 points); **// 0.3 points**
10. Implement Disabling test on condition (0.1 points); **// missing**
11. Test quality and adequate coverage will be assessed as (0.5 points).**// 0.5 points.**

kgebregewergs-epam committed:
Implemented ParameterizedTest for TemplateEngineTest

QwerTech commented:
2. @parameterized runner and Dynamic tests (0.3 points); **// 0.2 points now, dynamic tests are still missing**

kgebregewergs-epam committed:
Added test using a temporary folder
Added a meta annotation test that will run tests with a tag @UnitTest.

Refactored a method that will get input from file.

QwerTech commented:
TemporaryFolder rule (0.3 points); **// 0.3 points**

Implement meta-annotations and filtering (0.3 points); // 0.3 points
